File: services/reporting/src/reporting/application/ports.py
from __future__ import annotations

from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import Any, Protocol

from train_ticket_platform.events import EventEnvelope
from train_ticket_platform.messaging import HandlerResult, HandlerStatus, PublishFailed, SubscribeFailed

EventHandler = Callable[[EventEnvelope], HandlerResult]


class EventPublisher(Protocol):
    def publish(self, envelope: EventEnvelope) -> None:
        """Publish a fully-populated EventEnvelope."""


class EventSubscriber(Protocol):
    def subscribe(
        self,
        streams: Sequence[str],
        group: str,
        consumer_name: str,
        handler: EventHandler,
    ) -> None:
        """Subscribe to event streams as a consumer-group member."""

    def stop(self) -> None:
        """Stop background polling and finish in-flight processing."""

__all__ = [
    "EventEnvelope",
    "EventHandler",
    "EventPublisher",
    "EventSubscriber",
    "HandlerResult",
    "HandlerStatus",
    "PublishFailed",
    "SubscribeFailed",
]

File: services/supplier-catalog/internal/application/ports.go
package application

import (
	"context"

	kitmsg "github.com/trainticket/greenfield/platform/go-kit/messaging"
)

type EventEnvelope = kitmsg.EventEnvelope

// EventPublisher publishes fully-populated EventEnvelope values to the event bus.
type EventPublisher interface {
	Publish(ctx context.Context, envelope EventEnvelope) error
}

// EventHandler handles a deserialized EventEnvelope.
type EventHandler func(context.Context, EventEnvelope) error

// EventSubscriber subscribes to broker streams as a consumer group member.
type EventSubscriber interface {
	Subscribe(ctx context.Context, streams []string, group string, consumerName string, handler EventHandler) error
}
